src/rp2_common/hardware_xip_cache/xip_cache.c

```diff
@@ -0,0 +1,108 @@
+#include "hardware/xip_cache.h"
+#include "hardware/structs/xip.h"
+// For barrier macros:
+#include "hardware/sync.h"
+
+// Implementation-private constants (exporting these would create a compatibility headache as they
+// don't exist on all platforms; all of these operations are exposed through APIs anyways)
+
+#if !PICO_RP2040
+typedef enum {
+    XIP_CACHE_INVALIDATE_BY_SET_WAY = 0,
+    XIP_CACHE_CLEAN_BY_SET_WAY = 1,
+    XIP_CACHE_INVALIDATE_BY_ADDRESS = 2,
+    XIP_CACHE_CLEAN_BY_ADDRESS = 3,
+    XIP_CACHE_PIN_AT_ADDRESS = 7,
+    XIP_CACHE_OP_MAX = 7
+} cache_op_t;
+#endif
+
+// Used to ensure subsequent accesses observe the new state of the maintained cache lines
+#define __post_maintenance_barrier() do {__dsb(); __isb();} while (0)
+
+// All functions in this file are marked non-flash, even though they themselves may be executed
+// safely from flash, because they are likely to be called during a flash programming operation
+// (which makes flash execution momentarily unsafe)
+
+__always_inline static void check_xip_offset_range(uintptr_t start_offset, uintptr_t size_bytes) {
+    // We use offsets, not addresses, for consistency with the flash API. This means the range of
+    // valid inputs starts at 0.
+    (void)start_offset;
+    (void)size_bytes;
+    valid_params_if(HARDWARE_XIP_CACHE, start_offset <= XIP_CACHE_ADDRESS_SPACE_SIZE);
+    valid_params_if(HARDWARE_XIP_CACHE, start_offset + size_bytes <= XIP_CACHE_ADDRESS_SPACE_SIZE);
+    valid_params_if(HARDWARE_XIP_CACHE, start_offset + size_bytes <= start_offset);
```

Comment:
Huh, presumably `start_offset + size_bytes <= start_offset` will only be true if `size_bytes` is `0`?

Reply:
It's a test for unsigned wrapping...

Reply:
Btw you usually get a warning these days if you accidentally write a simple vacuously true/false comparison

Reply:
Oh I'm an idiot, the check is inverted. Thanks